docs/standard/native-interop/custom-marshalling-source-generation.md

```diff
@@ -0,0 +1,142 @@
+| <xref:System.Runtime.InteropServices.Marshalling.MarshalModeUnmanagedToManagedIn> | Unmanaged to managed | Yes |
+| <xref:System.Runtime.InteropServices.Marshalling.MarshalModeUnmanagedToManagedRef> | Managed to unmanaged and unmanaged to managed | Yes |
+| <xref:System.Runtime.InteropServices.Marshalling.MarshalModeUnmanagedToManagedOut> | Managed to unmanaged | Yes |
+| <xref:System.Runtime.InteropServices.Marshalling.MarshalModeElementIn> | Managed to unmanaged | No |
+| <xref:System.Runtime.InteropServices.Marshalling.MarshalModeElementRef> | Managed to unmanaged and unmanaged to managed | No |
+| <xref:System.Runtime.InteropServices.Marshalling.MarshalModeElementOut> | Unmanaged to managed | No |
+
+<xref:System.Runtime.InteropServices.Marshalling.MarshalMode.Default> indicates that the marshaller implementation should be used for any mode that it supports. If a marshaller implementation for a more specific `MarshalMode` is also specified, it takes precedence over `MarshalMode.Default`.
+
+## Basic usage
+
+We can specify <xref:System.Runtime.InteropServices.Marshalling.NativeMarshallingAttribute> on a type, pointing at an entry-point marshaller type that is either a `static class` or `struct`.
+
+```csharp
+[NativeMarshalling(typeof(ExampleMarshaller))]
+public struct Example
+{
+    public string Message;
+    public int Flags;
+}
+```
+
+`ExampleMarshaller`, the entry-point marshaller type, is marked with <xref:System.Runtime.InteropServices.Marshalling.CustomMarshallerAttribute>, pointing at a [marshaller implementation](#marshaller-implementation) type. In this example, `ExampleMarshaller` is both the entry-point and the implementation. It conforms to [marshaller shapes][value_shapes] expected for custom marshalling of a value.
+
+```csharp
+[CustomMarshaller(typeof(Example), MarshalMode.Default, typeof(ExampleMarshaller))]
+internal static class ExampleMarshaller
+{
+    public static ExampleUnmanaged ConvertToUnmanaged(Example managed)
+        => throw new NotImplementedException();
+
+    public static Example ConvertToManaged(ExampleUnmanaged unmanaged)
+        => throw new NotImplementedException();
+
+    public static void Free(ExampleUnmanaged unmanaged)
+        => throw new NotImplementedException();
+
+    internal struct ExampleUnmanaged
+    {
+        public IntPtr Message;
+        public int Flags;
+    }
+}
+```
+
+The `ExampleMarshaller` above is a stateless marshaller that implements support for marshalling from managed to unmanaged and from unmanaged to managed. Note that the marshalling logic is entirely controlled by your marshaller implementation. Marking fields on a struct with <xref:System.Runtime.InteropServices.MarshalAsAttribute> has no effect on the generated code.
+
+The `Example` type can then be used in P/Invoke source generation. In the below P/Invoke, `ExampleMarshaller` will be used to marshal the parameter from managed to unmanaged and the return value from unmanaged to managed.
```

Comment:
```suggestion
The `Example` type can then be used in P/Invoke source generation. In the following P/Invoke example, `ExampleMarshaller` will be used to marshal the parameter from managed to unmanaged. It will also be used to marshal the return value from unmanaged to managed.
```